A4.md

```diff
@@ -0,0 +1,266 @@
+
+The specification that tells us how to extract
+the headers will be configured as part of the [service
+config](https://github.com/grpc/grpc/blob/master/doc/service_config.md).
+There will also be a way to explicitly specify it via the gRPC client API.
+
+Note that for streaming RPCs, we extract headers only from the first
+request message on the stream.
+
+Note that because we need to see the first request message before we can
+determine the set of headers to send, we need to defer sending initial
+metadata until the client sends the first message, which is a change
+from the normal behavior.  This new behavior will take effect only if
+there is a header extraction config specified for the call.
+
+### Specification for Extracting Headers
+
+The specification of how to construct headers from the request payload
+will be configured via a JSON string of the following form:
+
+```
+// A list of zero or more header specifications.
+[
+  {
+    // The field name to extract from the payload.
+    // When protocol buffers are used for the payload, this must refer to
+    // a non-repeated string field.  If it is a field inside of another
+    // field (i.e., in a nested message), then '.' characters may be used
+    // to delimit the path elements, but none of the fields along that
+    // path may be repeated fields.  If the specified path does not exist
+    // in the payload, then no header will be added.
+    'payloadFieldName': string,
+
+    // Delimiter character, used both when splitting the value from the
+    // payload field into elements and when re-joining those elements to
+    // form the header value.  Must be a single ASCII character; if any
+    // other value is specified, then no header will be added.  If absent,
+    // the entire payload field will be used.
+    'delimiterCharacter': string,
+
+    // Number of split elements to keep (integer).
+    // Note that leading delimiter characters (those at the beginning of
+    // the payload field) are skipped; they do not count toward this
+    // number and are not added to the resulting header field.  If set
+    // to 0 or a negative value, no header will be added.  Ignored if the
+    // 'delimiterCharacter' field is not present.
+    'numElementsToKeep': number,
+
+    // Name of header to add.
+    // Note that if multiple entries list the same header name, or if
+    // the application already specified a header with this name, then
+    // the request will wind up with multiple headers of the same name.
+    'headerName': string
+  }
+]
+```
+
+When encoded in the service config, the above JSON object will be the
+value of a per-method parameter called `headerExtraction`.
+
+### Example
+
+Let's say that the request contains the following fields (protobuf syntax):
+
+```
+resource: {
+  id: "//foo/bar/baz"
+}
+user: "roth@quux@mumble@frotz"
+transaction_id: "some_big_opaque_string"
+```
+
+And let's say that we're using the following header extraction
```

Comment:
Micro nit: The 'we' feels a little too informal for a specification.  Also, it seems to change between 'we' the gRPC team, and 'we' gRPC plus the end user.  I personally find it distracting to the actual content.